### PR TITLE
Initial NI-FGEN add

### LINKED ISSUE #181: Add NI-FGEN

### PR DESCRIPTION
**Pull request by texasaggie97**
[X] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).
[X] I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.
~~[ ] I've added any unit tests that are applicable for this pull request~~
### What does this Pull Request accomplish?
* Initial support for NI-FGEN

### List issues fixed by this Pull Request below, if any.
* Fix #181 

### What testing has been done?
* Travis
* System tests

### CODE REVIEW

**Comment by marcoskirsch** [comment]
FGen -> NI-FGEN

File: CHANGELOG.md [hunk 1]
```diff
@@ -25,6 +25,11 @@ and this project adheres to [Python Versioning](http://legacy.python.org/dev/pep
  * #### Added
  * #### Changed
  * #### Removed
+* ### NI-FGen
```
Comment (marcoskirsch) [inline CHANGELOG.md]: NI-Fgen -> NI-FGEN

File: CHANGELOG.md [hunk 2]
```diff
@@ -104,5 +109,9 @@ and this project adheres to [Python Versioning](http://legacy.python.org/dev/pep
...
```
Comment (marcoskirsch) [inline CHANGELOG.md]: NI-Fgen -> NI-FGEN

File: build/helper/helper.py [hunk 3]
```diff
@@ -56,7 +56,10 @@ def get_python_type_for_visa_type(visa_type):
         'ViRsrc': 'str',
     }
 
-    return type_map[visa_type]
+    if visa_type in type_map:
+        return type_map[visa_type]
+    else:
+        return 'int'
```
Comment (marcoskirsch) [inline build/helper/helper.py]: When did you encounter this?
This catch-all makes me a little nervous.

File: src/nifgen/metadata/config.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+config = {
+    'metadata_version': '1.0',
+    'module_name': 'nifgen',
+    'module_version': '0.4.0.dev0',
+    'c_function_prefix': 'niFgen_',
+    'driver_name': 'NI-FGEN',
+    'session_class_description': 'An NI-FGEN session to a National Instruments Function Generator.',
```
Comment (marcoskirsch) [inline src/nifgen/metadata/config.py]: Function Generator -> Signal Generator

"function generators" is a subset of "signal generators", the other type of "signal generator" we sell being an "arbitrary waveform generator".

File: src/nifgen/metadata/functions_addon.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+# These dictionaries are applied to the generated functions dictionary at build time
+# Any changes to the API should be made here. functions.py is code generated
+
+# By default all functions in functions.py will be generated as a public function
+# This will override that with private - add '_' to the beginning of the name, or
+# don't generate at all
+functions_codegen_method = {
+    'InitWithOptions':                 { 'codegen_method': 'private',       },
```
Comment (marcoskirsch) [inline src/nifgen/metadata/functions_addon.py]: minor - all other } align

File: src/nifgen/metadata/functions_addon.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
+# These dictionaries are applied to the generated functions dictionary at build time
+# Any changes to the API should be made here. functions.py is code generated
+
+# By default all functions in functions.py will be generated as a public function
+# This will override that with private - add '_' to the beginning of the name, or
+# don't generate at all
+functions_codegen_method = {
```
Comment (marcoskirsch) [inline src/nifgen/metadata/functions_addon.py]: For a separate PR (you may want to add a TODO) there's still a good amount of functions to add here, I think.

File: build/helper/helper.py (hunk 3 above)
Comment (texasaggie97) [reply build/helper/helper.py]: FGEN has NIComplexNumber and NIComplexI16. I was not planning on supporting custom (non-visa) types in this PR.

File: CHANGELOG.md (hunk 1 above)
Comment (texasaggie97) [reply CHANGELOG.md]: Fixed

File: CHANGELOG.md (hunk 2 above)
Comment (texasaggie97) [reply CHANGELOG.md]: Fixed

File: src/nifgen/metadata/config.py (hunk 4 above)
Comment (texasaggie97) [reply src/nifgen/metadata/config.py]: Fixed

File: src/nifgen/metadata/functions_addon.py (hunk 6 above)
Comment (texasaggie97) [reply src/nifgen/metadata/functions_addon.py]: TODO added

File: src/nifgen/metadata/functions_addon.py (hunk 5 above)
Comment (texasaggie97) [reply src/nifgen/metadata/functions_addon.py]: Fixed

File: build/helper/helper.py (hunk 3 above)
Comment (marcoskirsch) [reply build/helper/helper.py]: Can you return None or '' or something obviously invalid?
We should not be generating code for these functions yet anyway.

Comment (texasaggie97) [reply build/helper/helper.py]: Not without putting workarounds elsewhere that I don't want to do just to make this work.

This way, all the templates generate SOMETHING. That something may not be correct for the DLL though. That will need custom types (or some other mechanism)

Comment (marcoskirsch) [reply build/helper/helper.py]: > That something may not be correct for the DLL though. 

This is what worries me, silently generating code that calls into the DLL incorrectly, corrupting memory or crashing. If we simply assert "type not supported" but _only_ when the function is to be generated, that may be enough. Workaround is to temporarily disable code generation.

**Comment by marcoskirsch** [comment]
In src/nifgen/metadata/functions.py it references niFgen\_WriteNamedWaveformComplexi16 (not lowercase i). It should be WriteNamedWaveformComplexI16. This is a minor problem likely when the metadata is extracted using NI's internal tools.

**Comment by marcoskirsch** [comment]
> Not without putting workarounds elsewhere that I don't want to do just to make this work.

I did a little bit of digging to understand this better.

I skip functions that aren't included in the Python API (codegen-method == 'no') during the add_all_function_metadata() step. Then I added all functions with Complex in their name to functions_add_on.py so they don't get generated. I also had to include 'GetStreamEndpointHandle' (used peer-to-peer) because it uses void* !

Once I did this, NI-FGEN generates successfully.

A side effect of the add_all_function_metadata() change is that while generating RST documentation, we get a warning in the form "Warning: "MyFunction" not found in function metadata. Typo? Generated code will be funky!" whenever creating what will be a broken link to said function. Which is a good thing?

**Comment by texasaggie97** [comment]
This is an issue in the source. I’ve started a list of problems there but so far, this has been the only thing.

[RFMI_Puzzle_Logo_cubed_Sm]

Mark Silva
National Instruments | Senior Software Engineer

(512) 683-8239 Work | mark.silva@ni.com<mailto:mark.silva@ni.com>
http://niweb.natinst.com/confluence/x/3KGQG

**Pull request merged**